src/censored.jl

```diff
@@ -0,0 +1,519 @@
+    log_prob_trunc = log1mexp(log_prob_upper)
+    prob_trunc = exp(log_prob_trunc)
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(log_prob_trunc)) && return entropy_bound
+    dtrunc = _to_truncated(d)
+    entropy_interval = prob_trunc * entropy(dtrunc) - xexpx(log_prob_trunc) + xlogx_pu
+    return oftype(entropy_bound, entropy_bound + entropy_interval)
+end
+function entropy(d::RightCensored)
+    d0 = d.uncensored
+    lower = d.lower
+    log_prob_lower_inc = logcdf(d0, lower)
+    if value_support(typeof(d0)) === Discrete
+        logpl = logpdf(d0, lower)
+        log_prob_lower = logsubexp(log_prob_lower_inc, logpl)
+        xlogx_pl = xexpx(logpl)
+    else
+        log_prob_lower = log_prob_lower_inc
+        xlogx_pl = 0
+    end
+    entropy_bound = -xexpx(log_prob_lower_inc)
+    log_prob_trunc = log1mexp(log_prob_lower)
+    prob_trunc = exp(log_prob_trunc)
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(log_prob_trunc)) && return entropy_bound
+    dtrunc = _to_truncated(d)
+    entropy_interval = prob_trunc * entropy(dtrunc) - xexpx(log_prob_trunc) + xlogx_pl
+    return oftype(entropy_bound, entropy_bound + entropy_interval)
+end
+function entropy(d::Censored)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    log_prob_lower_inc = logcdf(d0, lower)
+    log_prob_upper = logccdf(d0, upper)
+    if value_support(typeof(d0)) === Discrete
+        logpl = logpdf(d0, lower)
+        logpu = logpdf(d0, upper)
+        log_prob_lower = logsubexp(log_prob_lower_inc, logpl)
+        log_prob_upper_inc = logaddexp(log_prob_upper, logpu)
+        xlogx_pl = xexpx(logpl)
+        xlogx_pu = xexpx(logpu)
+    else
+        log_prob_lower = log_prob_lower_inc
+        log_prob_upper_inc = log_prob_upper
+        xlogx_pl = xlogx_pu = 0
+    end
+    entropy_bound = -(xexpx(log_prob_lower_inc) + xexpx(log_prob_upper_inc))
+    log_prob_trunc = log1mexp(logaddexp(log_prob_lower, log_prob_upper))    
+    prob_trunc = exp(log_prob_trunc)
+    # truncation contains ~ no probability
+    prob_trunc < eps(one(log_prob_trunc)) && return entropy_bound
+    dtrunc = _to_truncated(d)
+    entropy_interval = prob_trunc * entropy(dtrunc) - xexpx(log_prob_trunc) + xlogx_pl + xlogx_pu
+    return entropy_bound + entropy_interval
+end
+
+
+#### Evaluation
+
+function pdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    px = float(pdf(d0, x))
+    return if _in_open_interval(x, lower, upper)
+        px
+    elseif x == lower
+        x == upper ? one(px) : oftype(px, cdf(d0, x))
+    elseif x == upper
+        if value_support(typeof(d0)) === Discrete
+            oftype(px, ccdf(d0, x) + px)
+        else
+            oftype(px, ccdf(d0, x))
+        end
+    else  # not in support
+        zero(px)
+    end
+end
+
+function logpdf(d::Censored, x::Real)
+    d0 = d.uncensored
+    lower = d.lower
+    upper = d.upper
+    logpx = float(logpdf(d0, x))
```

Comment:
We assume throughout Distributions that `logpdf` returns something that `-Inf` can be converted to, i.e., it's already a floating point number, so `float` shouldn't be needed here either:
 ```suggestion
     logpx = logpdf(d0, x)
```

Reply:
ah, good to know